ui/app/components/auth-jwt.js

```diff
@@ -0,0 +1,165 @@
+    return this.window || window;
+  },
+
+  fetchRole: task(function*(roleName, options = { debounce: true }) {
+    if (options.debounce) {
+      this.onRoleName(roleName);
+      // debounce
+      yield timeout(WAIT_TIME);
+    }
+    let path = this.selectedAuthPath || 'jwt';
+    let id = JSON.stringify([path, roleName]);
+    let role = null;
+    try {
+      role = yield this.store.findRecord('role-jwt', id, { adapterOptions: { namespace: this.namespace } });
+    } catch (e) {
+      if (!e.httpStatus || e.httpStatus !== 400) {
+        throw e;
+      }
+    }
+    this.set('role', role);
+  }).restartable(),
+
+  handleOIDCError(err) {
+    this.onLoading(false);
+    this.prepareForOIDC.cancelAll();
+    this.onError(err);
+  },
+
+  prepareForOIDC: task(function*(oidcWindow) {
+    // show the loading animation in the parent
+    this.onLoading(true);
+    // start watching the popup window and the current one
+    this.watchPopup.perform(oidcWindow);
+    this.watchCurrent.perform(oidcWindow);
+    // and then wait for storage event to be fired from the popup
+    // window setting a value in localStorage when the callback route is loaded
+    let storageEvent = yield waitForEvent(this.getWindow(), 'storage');
+    this.exchangeOIDC.perform(storageEvent, oidcWindow);
+  }),
+
+  watchPopup: task(function*(oidcWindow) {
+    while (true) {
+      yield timeout(WAIT_TIME);
+      if (!oidcWindow || oidcWindow.closed) {
+        return this.handleOIDCError(ERROR_WINDOW_CLOSED);
+      }
+    }
+  }),
+
+  watchCurrent: task(function*(oidcWindow) {
+    yield waitForEvent(this.getWindow(), 'beforeunload');
+    oidcWindow.close();
+  }),
+
+  closeWindow(oidcWindow) {
+    this.watchPopup.cancelAll();
+    this.watchCurrent.cancelAll();
+    oidcWindow.close();
+  },
+
+  exchangeOIDC: task(function*(event, oidcWindow) {
+    if (event.key !== 'oidcState') {
+      return;
+    }
+    this.onLoading(true);
+    // get the info from the event fired by the other window and
+    // then remove it from localStorage
+    let { namespace, path, state, code } = JSON.parse(event.newValue);
+    this.getWindow().localStorage.removeItem('oidcState');
+
+    // defer closing of the window, but continue executing the task
+    later(() => {
+      this.closeWindow(oidcWindow);
+    }, WAIT_TIME);
+    if (!path || !state || !code) {
+      return this.handleOIDCError(ERROR_MISSING_PARAMS);
+    }
+    let adapter = this.store.adapterFor('auth-method');
+    this.onNamespace(namespace);
+    let resp;
+    // do the OIDC exchange, set the token on the parent component
+    // and submit auth form
+    try {
+      resp = yield adapter.exchangeOIDC(path, state, code);
+    } catch (e) {
+      return this.handleOIDCError(e);
+    }
+    let token = resp.auth.client_token;
+    this.onSelectedAuth('token');
+    this.onToken(token);
+    yield this.onSubmit();
+  }),
+
+  actions: {
+    async startOIDCAuth(data, e) {
+      this.onError(null);
```

Comment:
what does calling `onError` here do?

Reply:
`onError` is an action passed from the outside here: https://github.com/hashicorp/vault/pull/6188/files/7456c6c0bf7f46f65d2fdcc47b1b3765ffc99d53#diff-5f637ddcdf2b0444777d3addb777746cR59 - the `mut this.error` is a shorthand syntax for setting that property so anytime `this.onError` is called in auth-jwt, the argument we pass will be set as the value of `error` on `auth-form`. Here, passing `null` just clears out any error when a user clicks the button to start the flow.